### ISSUE TITLE
Add general switch to customize docker daemon port

### ISSUE THREAD

**Issue by the42** [docker/machine#2820]
The Azure-driver provides the ability to specify the port on which docker, after host creation, should listen
    https://docs.docker.com/machine/drivers/azure/
--azure-docker-port

However, I definitely think that this switch need not be restricted to a driver but be provided as a general switch.

Rationale: provider uses eg. openstack but does not allow communication on port 2376

**Comment by nathanleclaire** [comment]
Thanks for the issue @the42. It's definitely something we could implement at the provisioner level, _but_ the really tricky bit is modifying all of the drivers to open the security group ports (most open `:2376` by default) properly.

**Comment by the42** [comment]
In that case I was misunderstood and/or didn't read clear the switch --azure-docker-port.
Opening ports in a firewall clearly is a driver issue. What I am referring to is, what if a user managed to specify a port to use the docker service other than 2376 / 3376 (because those ports are unavailable on the hosting provider for what reasons ever), it is still required that a  command like
    docker-machine ls 
or
    docker-machine env <machine-name>

connects to that specific port. So it should be possible to specify that port during docker-machine create and therefore need not necessarily be a driver supported feature.

Right, if docker-machine create should also support opening ports in FW-rules, it has to be driver specific.

**Comment by ahmetalpbalkan** [comment]
I can also see that most drivers hardcode :2376. @the42 do you have an Azure-specific problem with this port number?

**Comment by the42** [comment]
@ahmetalpbalkan no, actually on an OpenStack implementation where port 2376 is closed and docker has to run on a different port. As I am unable to configure docker to run on a different port via docker-machine (or I am unaware how to do that) I am stuck.

**Comment by jamshid** [comment]
I really want this capability too, is there a variable I can set in ~/.docker/machine/machines/MACHINE/config.json to override the docker port for when it's *not* 2376?

In my case I'm SSH-proxying multiple docker servers -- two ports each, docker 2376 and ssh 22. I need to do this because I do not have direct access to the docker servers when remote.

When I'm remote I can simply change IPAddress  and SSHPort in config.json to the proxied port, but need a config option such that DOCKER_HOST gets set to tcp://127.0.0.1:<docker-proxy-port>.

This would allow me to use the regular docker-machine workflow for both local and remote access. Of course, docker-machine doesn't actually run the ssh proxies -- that'd be cool though! 

Maybe this scenario is already covered by swarm or I should learn how to setup a VPN.

**Comment by dweomer** [comment]
:+1:

**Comment by fsoppelsa** [comment]
Maybe `config.json` would be a good start point to store the port.

**Comment by ahmetalpbalkan** [comment]
BTW The new azure driver has a `--azure-docker-port` option as well.

It just seems like someone should take initiative to make this port common enough for all drivers by making it an option parsed into the BaseDriver, just like the `--swarm-host` argument. It's doable (though it'll require more changes to cloud drivers)

**Comment by fsoppelsa** [comment]
@ahmetalpbalkan I can take this in case.

**Comment by jonasrauber** [comment]
Any progress?

**Comment by jonasrauber** [comment]
Maybe we can at least add an option to specific drivers (in particular the AWS driver), similar to the Azure driver.
